public/components/hooks/use_polling.ts

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { useState, useEffect, useRef } from 'react';
+
+interface UsePollingProps<T> {
+  fetchFunction: () => Promise<T>;
+  initialInterval?: number;
+  cachingStrategy?: (item: T, cache: Record<string, T>) => void;
+  enableExponentialBackoff?: boolean;
+  maxRetries?: number;
+  maxInterval?: number;
+}
+
+interface UsePollingResponse<T> {
+  data: T | null;
+  cache: Record<string, T>;
+  startPolling: () => void;
+  stopPolling: () => void;
+  setInterval: (interval: number) => void;
+  isPolling: boolean;
+  error: any;
+  interval: number;
+}
+
+/**
+ * usePolling - A hook for polling data asynchronously at regular intervals.
+ *
+ * @template T - The type of data expected to be fetched and managed by the hook.
+ * @param {UsePollingProps<T>} props - The configuration options for the hook.
+ * @returns {UsePollingResponse<T>} - An object containing the polled data, cache, control functions, and other states.
+ */
+
+export function usePolling<T>({
+  fetchFunction,
+  initialInterval = 1000,
+  cachingStrategy,
+  enableExponentialBackoff = false,
+  maxRetries = 10, // Add a limit to the number of retries to prevent infinite retries
+  maxInterval = 60000, // Add a maximum interval, e.g., 60 seconds, to ensure the interval doesn't become unmanageably large
+}: UsePollingProps<T>): UsePollingResponse<T> {
+  const [data, setData] = useState<T | null>(null); // The latest fetched data
```

Comment:
curious why null instead of undefined? `data` shouldn't be defined initially, same for `error`